src/nn/embedding.rs

```diff
@@ -0,0 +1,206 @@
+use crate::{optim::*, shapes::*, tensor::*, tensor_ops::*};
+
+use super::module::{Module, ModuleMut, ResetParams};
+
+/// An embedding
+/// Initializes [Self::weight] from a Uniform distribution
+/// between [-1 / sqrt(I), 1 / sqrt(I)].
+///
+/// # Generics
+/// - `VOCAB` The size of the vocabulary, inputs integer values must be between
+///    0 and VOCAB;
+/// - `DIM` The "output" size of vectors & matrices which are the vectors being selected.
+///
+/// # Examples
+/// `Embedding<5, 2>` can act on vectors with SEQ integer elements (with values between 0 and 4), and results in a SEQ tensor of
+/// usually f32 elements being the rows in [Self::weight].
+/// ```rust
+///
+/// # use dfdx::prelude::*;
+/// # let dev: Cpu = Default::default();
+/// let model: Embedding<5, 2> = dev.build_module();
+/// // single item forward
+/// let inputs: Tensor<(Const<5>,), usize> = dev.zeros::<Rank1<5>>();
+/// let _: Tensor<(Const<2>,), f32> = model.forward(inputs);
+/// // batched forward
+/// let inputs: Tensor<(Const<10>, Const<5>,), usize> = dev.zeros::<Rank2<10, 5>>();
+/// let _: Tensor<(Const<10>, Const<2>), f32> = model.forward(inputs);
+/// ```
+#[derive(Debug, Clone)]
+pub struct Embedding<const VOCAB: usize, const DIM: usize, D: Device<f32> = Cpu> {
+    /// Transposed weight matrix, shape (I, O)
+    pub weight: Tensor<Rank2<VOCAB, DIM>, f32, D>,
+}
+
+impl<const VOCAB: usize, const DIM: usize, const SEQ: usize, D: Device<f32>>
+    Module<Tensor<Rank1<SEQ>, usize, D>> for Embedding<VOCAB, DIM, D>
+{
+    type Output = Tensor<Rank2<SEQ, DIM>, f32, D>;
+    fn forward(&self, input: Tensor<Rank1<SEQ>, usize, D>) -> Self::Output {
+        self.weight.retaped().gather(input)
```

Comment:
This is the particular line that seems to be not working.

In other layer I found

```rust
self.weight.retaped::<T>()  // T is the tape of the input 
```
Which seems to be the trick. However, it seems to me that the weight could contain the tape, and the input cannot ( since it's only indexing within the weight tensor).

I'm out of ideas to make this work. Could you provide any help @coreylowman  ?

Reply:
I figure out a way by making `GatherTo` generic on the Tape too. I'm not to sure about the modifications though.